File: examples/kotlin/cases/tasks/task_move/task_move.domain.case.kt
package app.protocol.examples.kotlin.cases.tasks.task_move

import app.protocol.examples.kotlin.core.AppSchema
import app.protocol.examples.kotlin.core.BaseDomainCase
import app.protocol.examples.kotlin.core.DomainExample
import kotlinx.serialization.Serializable

private val taskStatusValues = listOf("todo", "doing", "done")

@Serializable
data class Task(
    val id: String,
    val title: String,
    val description: String? = null,
    val status: String,
    val createdAt: String,
    val updatedAt: String,
)

@Serializable
data class TaskMoveInput(
    val taskId: String,
    val targetStatus: String,
)

@Serializable
data class TaskMoveOutput(
    val task: Task,
)

fun assertTaskRecord(task: Task, source: String) {
    require(task.id.isNotBlank()) { "$source.id must be a non-empty string" }
    require(task.title.isNotBlank()) { "$source.title must be a non-empty string" }
    require(task.createdAt.isNotBlank()) { "$source.createdAt must be a non-empty string" }
    require(task.updatedAt.isNotBlank()) { "$source.updatedAt must be a non-empty string" }
    require(task.status in taskStatusValues) {
        "$source.status must be one of ${taskStatusValues.joinToString()}"
    }
}

class TaskMoveDomain : BaseDomainCase<TaskMoveInput, TaskMoveOutput>() {
    override fun caseName(): String = "task_move"

    override fun description(): String =
        "Moves an existing task card to another board column."

    override fun inputSchema(): AppSchema =
        AppSchema(
            type = "object",
            properties = mapOf(
                "taskId" to AppSchema(
                    type = "string",
                    description = "Identifier of the task that will be moved.",
                ),
                "targetStatus" to AppSchema(
                    type = "string",
                    description = "Destination board column for the task.",
                    enum = taskStatusValues,
                ),
            ),
            required = listOf("taskId", "targetStatus"),
            additionalProperties = false,
        )

    override fun outputSchema(): AppSchema =
        AppSchema(
            type = "object",
            properties = mapOf(
                "task" to AppSchema(
                    type = "object",
                    properties = mapOf(
                        "id" to AppSchema(type = "string"),
                        "title" to AppSchema(type = "string"),
                        "description" to AppSchema(type = "string"),
                        "status" to AppSchema(type = "string", enum = taskStatusValues),
                        "createdAt" to AppSchema(type = "string"),
                        "updatedAt" to AppSchema(type = "string"),
                    ),
                    required = listOf("id", "title", "status", "createdAt", "updatedAt"),
                    additionalProperties = false,
                ),
            ),
            required = listOf("task"),
            additionalProperties = false,
        )

    override fun validate(input: TaskMoveInput) {
        require(input.taskId.isNotBlank()) { "taskId is required" }
        require(input.targetStatus in taskStatusValues) {
            "targetStatus must be one of ${taskStatusValues.joinToString()}"
        }
    }

    fun validateOutput(output: TaskMoveOutput) {
        assertTaskRecord(output.task, "task_move.output.task")
    }

    override fun invariants(): List<String> =
        listOf(
            "Moving a task never changes its identity.",
            "A move only updates status and, when applicable, updatedAt.",
            "Moving to the same status is idempotent and returns the unchanged task.",
        )

    override fun examples(): List<DomainExample<TaskMoveInput, TaskMoveOutput>> =
        listOf(
            DomainExample(
                name = "move_todo_to_doing",
                description = "A task leaves todo and enters doing.",
                input = TaskMoveInput(taskId = "task_001", targetStatus = "doing"),
                output = TaskMoveOutput(
                    task = Task(
                        id = "task_001",
                        title = "Ship the Kotlin example",
                        status = "doing",
                        createdAt = "2026-03-18T12:00:00Z",
                        updatedAt = "2026-03-18T12:20:00Z",
                    ),
                ),
            ),
            DomainExample(
                name = "idempotent_move",
                description = "Moving to the same status keeps the task unchanged.",
                input = TaskMoveInput(taskId = "task_002", targetStatus = "done"),
                output = TaskMoveOutput(
                    task = Task(
                        id = "task_002",
                        title = "Prepare release notes",
                        status = "done",
                        createdAt = "2026-03-18T12:10:00Z",
                        updatedAt = "2026-03-18T12:10:00Z",
                    ),
                ),
            ),
        )

    override suspend fun test() {
        validate(TaskMoveInput(taskId = "task_001", targetStatus = "doing"))

        check(
            runCatching { validate(TaskMoveInput(taskId = "", targetStatus = "doing")) }.isFailure,
        ) {
            "test: validate must reject empty taskId"
        }

        check(
            runCatching { validate(TaskMoveInput(taskId = "task_001", targetStatus = "invalid")) }.isFailure,
        ) {
            "test: validate must reject invalid targetStatus"
        }

        validateOutput(
            TaskMoveOutput(
                task = Task(
                    id = "task_001",
                    title = "Valid task",
                    status = "doing",
                    createdAt = "2026-03-18T12:00:00Z",
                    updatedAt = "2026-03-18T12:20:00Z",
                ),
            ),
        )
    }
}

File: examples/node/cases/tasks/task_move/task_move.domain.case.ts
import {
  AppSchema,
  BaseDomainCase,
  type DomainExample,
} from "../../../core/domain.case";

export type TaskStatus = "todo" | "doing" | "done";

export interface Task {
  id: string;
  title: string;
  description?: string;
  status: TaskStatus;
  createdAt: string;
  updatedAt: string;
}

export interface TaskMoveInput {
  taskId: string;
  targetStatus: TaskStatus;
}

export interface TaskMoveOutput {
  task: Task;
}

const TASK_STATUS_VALUES: TaskStatus[] = ["todo", "doing", "done"];

function isTaskStatus(value: unknown): value is TaskStatus {
  return (
    typeof value === "string" &&
    TASK_STATUS_VALUES.includes(value as TaskStatus)
  );
}

export function assertTaskRecord(
  value: unknown,
  source: string
): asserts value is Task {
  if (typeof value !== "object" || value === null) {
    throw new Error(`${source} must be an object`);
  }

  const record = value as Record<string, unknown>;
  const requiredStringFields = ["id", "title", "createdAt", "updatedAt"] as const;

  for (const field of requiredStringFields) {
    if (
      typeof record[field] !== "string" ||
      (record[field] as string).trim().length === 0
    ) {
      throw new Error(`${source}.${field} must be a non-empty string`);
    }
  }

  if (!isTaskStatus(record.status)) {
    throw new Error(
      `${source}.status must be one of ${TASK_STATUS_VALUES.join(", ")}`
    );
  }

  if (
    record.description !== undefined &&
    typeof record.description !== "string"
  ) {
    throw new Error(`${source}.description must be a string when provided`);
  }
}

export class TaskMoveDomain extends BaseDomainCase<
  TaskMoveInput,
  TaskMoveOutput
> {
  public caseName(): string {
    return "task_move";
  }

  public description(): string {
    return "Moves an existing task card to another board column.";
  }

  public inputSchema(): AppSchema {
    return {
      type: "object",
      properties: {
        taskId: {
          type: "string",
          description: "Identifier of the task that will be moved.",
        },
        targetStatus: {
          type: "string",
          description: "Destination board column for the task.",
          enum: [...TASK_STATUS_VALUES],
        },
      },
      required: ["taskId", "targetStatus"],
      additionalProperties: false,
    };
  }

  public outputSchema(): AppSchema {
    return {
      type: "object",
      properties: {
        task: {
          type: "object",
          properties: {
            id: { type: "string" },
            title: { type: "string" },
            description: { type: "string" },
            status: { type: "string", enum: [...TASK_STATUS_VALUES] },
            createdAt: { type: "string" },
            updatedAt: { type: "string" },
          },
          required: ["id", "title", "status", "createdAt", "updatedAt"],
          additionalProperties: false,
        },
      },
      required: ["task"],
      additionalProperties: false,
    };
  }

  public validate(input: TaskMoveInput): void {
    if (typeof input !== "object" || input === null) {
      throw new Error("input must be an object");
    }

    if (typeof input.taskId !== "string" || input.taskId.trim().length === 0) {
      throw new Error("taskId is required");
    }

    if (!isTaskStatus(input.targetStatus)) {
      throw new Error(
        `targetStatus must be one of ${TASK_STATUS_VALUES.join(", ")}`
      );
    }
  }

  public validateOutput(output: TaskMoveOutput): void {
    if (typeof output !== "object" || output === null) {
      throw new Error("output must be an object");
    }

    assertTaskRecord(output.task, "task_move.output.task");
  }

  public invariants(): string[] {
    return [
      "Moving a task never changes its identity.",
      "A move only updates status and, when applicable, updatedAt.",
      "Moving to the same status is idempotent and returns the unchanged task.",
    ];
  }

  public examples(): DomainExample<TaskMoveInput, TaskMoveOutput>[] {
    return [
      {
        name: "move_todo_to_doing",
        description: "A task leaves todo and enters doing.",
        input: {
          taskId: "task_001",
          targetStatus: "doing",
        },
        output: {
          task: {
            id: "task_001",
            title: "Ship the Node example",
            status: "doing",
            createdAt: "2026-03-18T12:00:00.000Z",
            updatedAt: "2026-03-18T12:20:00.000Z",
          },
        },
      },
      {
        name: "idempotent_move",
        description: "Moving to the same status keeps the task unchanged.",
        input: {
          taskId: "task_002",
          targetStatus: "done",
        },
        output: {
          task: {
            id: "task_002",
            title: "Prepare release notes",
            status: "done",
            createdAt: "2026-03-18T12:10:00.000Z",
            updatedAt: "2026-03-18T12:10:00.000Z",
          },
        },
      },
    ];
  }

  public async test(): Promise<void> {
    this.validate({
      taskId: "task_001",
      targetStatus: "doing",
    });

    let threw = false;
    try {
      this.validate({
        taskId: "",
        targetStatus: "doing",
      });
    } catch {
      threw = true;
    }

    if (!threw) {
      throw new Error("test: validate must reject empty taskId");
    }

    threw = false;
    try {
      this.validate({
        taskId: "task_001",
        targetStatus: "invalid" as TaskStatus,
      });
    } catch {
      threw = true;
    }

    if (!threw) {
      throw new Error("test: validate must reject invalid targetStatus");
    }

    this.validateOutput({
      task: {
        id: "task_001",
        title: "Valid task",
        status: "doing",
        createdAt: "2026-03-18T12:00:00.000Z",
        updatedAt: "2026-03-18T12:20:00.000Z",
      },
    });

    threw = false;
    try {
      this.validateOutput({
        task: {
          id: "task_001",
          title: "Broken task",
          description: 123 as unknown as string,
          status: "doing",
          createdAt: "2026-03-18T12:00:00.000Z",
          updatedAt: "2026-03-18T12:20:00.000Z",
        },
      });
    } catch {
      threw = true;
    }

    if (!threw) {
      throw new Error("test: validateOutput must reject invalid task payloads");
    }
  }
}
